File: backend/app/join_token.py
"""实时会话 join token(M1-B)—— HMAC 签名,stdlib 实现,无第三方依赖。

跨栈契约(已冻结,实时会话服务/bridge 侧对称实现,勿改格式):
  token = "v1.<session_id>.<exp_unix>.<sig>"
  sig   = base64url 无 padding( HMAC-SHA256(key=AIM_BRIDGE_CALLBACK_SECRET, msg="v1.<session_id>.<exp_unix>") )
  exp_unix = 秒级 Unix 时间戳。

客户端流程:GET /api/sessions/{id}/join → {join_token, ws_path, expires_at}
→ 连 wss://<站点>/rt/ws?session_id=<id> → 首帧 {"type":"auth","token":<join_token>}。

签发在控制面(routers/sessions.py);生产验签在实时会话服务(bridge)。
本模块的 verify_join_token 供单测对拍 + 契约钉死,与 bridge 侧逐字节等价。
"""
from __future__ import annotations

import base64
import hashlib
import hmac


def _b64url_nopad(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _sig(msg: str, secret: str) -> str:
    mac = hmac.new(secret.encode("utf-8"), msg.encode("utf-8"), hashlib.sha256)
    return _b64url_nopad(mac.digest())


def sign_join_token(session_id: str, exp_unix: int, secret: str) -> str:
    """签发 join token(契约见模块 docstring)。secret 空 = 配置缺失,调用方应先 fail-closed(503)。"""
    if not secret:
        raise ValueError("join token 密钥未配置(AIM_BRIDGE_CALLBACK_SECRET)")
    msg = f"v1.{session_id}.{int(exp_unix)}"
    return f"{msg}.{_sig(msg, secret)}"


def verify_join_token(token: str, secret: str, now_unix: int) -> str | None:
    """验签 + 校验有效期(fail-closed)。成功返回 session_id,否则 None(格式/签名/过期一律 None)。

    - 签名对拍用 token 里的**字面** exp 串重算(不经 int 归一,防 "07" 之类规范化歧义)。
    - hmac.compare_digest 常量时间比较防时序侧信道。
    - 过期判定与 candidate_token 同口径:now_unix > exp 即失效(恰好等于 exp 仍有效)。
    """
    if not token or not secret:
        return None
    parts = token.split(".")
    if len(parts) != 4 or parts[0] != "v1":
        return None
    _, session_id, exp_str, sig = parts
    if not session_id:
        return None
    if not hmac.compare_digest(sig, _sig(f"v1.{session_id}.{exp_str}", secret)):
        return None
    try:
        exp = int(exp_str)
    except ValueError:
        return None  # 签名虽对但 exp 非契约形态(秒级整数)→ 拒
    if int(now_unix) > exp:
        return None  # 过期
    return session_id

File: bridge/src/join-token.ts
/**
 * join token 验签(M1-B,跨栈契约冻结)。
 *
 * 格式:`v1.<session_id>.<exp_unix>.<sig>`
 *   sig = base64url 无 padding( HMAC-SHA256(key = AIM_BRIDGE_CALLBACK_SECRET 的值,
 *                                            msg = "v1.<session_id>.<exp_unix>") )
 *   exp_unix = 秒级 Unix 时间戳(nowMs >= exp*1000 即过期)。
 *
 * backend 侧签发(同一密钥双用途:/sessions/:id/ready 的 X-Bridge-Secret 与 join token HMAC key)。
 * 本模块只做纯函数验签:格式 / HMAC(timingSafeEqual 防时序侧信道)/ 过期。
 * fail-closed:secret 空 → 一律 null(未配密钥不得放行任何 token)。
 */
import { createHmac, timingSafeEqual } from "node:crypto";

/** token 长度上限(防超长输入进 HMAC/split;正常 token 远小于此)。 */
const MAX_TOKEN_LEN = 4096;

/**
 * 验签一枚 join token。
 * @returns 验签通过 → `{ sessionId }`;格式坏 / 签名不符 / 已过期 / secret 空 → `null`。
 */
export function verifyJoinToken(
  token: string,
  secret: string,
  nowMs: number = Date.now(),
): { sessionId: string } | null {
  if (!secret) return null; // fail-closed:密钥未配 → 拒绝一切 token
  if (typeof token !== "string" || token.length === 0 || token.length > MAX_TOKEN_LEN) return null;
  const parts = token.split(".");
  if (parts.length !== 4) return null;
  const [version, sessionId, expStr, sig] = parts;
  if (version !== "v1") return null;
  if (!sessionId) return null;
  if (!/^\d{1,15}$/.test(expStr)) return null; // 秒级时间戳:纯数字、防超长
  const exp = Number(expStr);
  if (!Number.isSafeInteger(exp)) return null;
  // 先验签后验期:两者失败对调用方等价(都是 auth_failed),顺序不泄漏信息。
  const expected = createHmac("sha256", secret).update(`v1.${sessionId}.${expStr}`).digest();
  const actual = Buffer.from(sig, "base64url");
  if (actual.length !== expected.length) return null;
  if (!timingSafeEqual(actual, expected)) return null;
  if (nowMs > exp * 1000) return null; // 过期(与 backend join_token.py 同口径:now > exp 才失效,恰等于 exp 仍有效)
  return { sessionId };
}
